`ot_uart`: Add break interrupt and add opt-in mocking of break oversampling

AlexJones0 opened the pull request:
This PR hooks up the QEMU UART RX break event to the `ot_uart` so that the break receive triggers an interrupt. This completely ignores the configured [`CTRL.RXBLVL`](https://opentitan.org/book/hw/ip/uart/doc/registers.html#ctrl--rxblvl) as we have no notion for "how long" break is being held for. Breaks can be sent to a given CharDev via the QEMU monitor.

With the goal of mocking the UART oversampling `VAL` register enough that it can support differentiation between break conditions and not (essentially treating the UART RX like a GPIO strap, as Earlgrey's ROM_EXT Rescue 1.0 protocol unfortunately does), we implement logic for outputting either `0x0` or `0xFFFF` depending on whether we are "in a break" or not. That is, the `VAL` register will report 16 samples equivalent to the last UART bit received (which is 0 when in a break, and 1 for a stop bit on any valid transmission). We make the assumption here that break is being "held", rather than just being a transient event. This logic is enabled through an `oversample-break` property to make it an opt-in behaviour.

To better support the ROM_EXT flows that use break like a strapping, add an additional property that treats incoming break events as a "toggle", such that one break event will cause the UART to enter break until another event tells it to exit break. This is similarly an opt-in property intended for testing edge-case software like ROM_EXT that uses UART Break in this way.

AlexJones0 committed:
[ot] hw/opentitan: ot_uart: Add break interrupt & mocked oversampling

This commit hooks up the QEMU UART RX break event to the `ot_uart` so
that the break receive triggers an interrupt. This completely ignores
the configured `CTRL.RXBLVL` as we have no notion for "how long" break
is being held for. Breaks can be sent to a given CharDev via the QEMU
monitor.

With the goal of mocking the UART oversampling `VAL` register enough
that it can support differentiation between break conditions and not
(essentially treating the UART RX like a GPIO strap), we implement
logic for outputting either 0x0 or 0xFFFF depending on whether we are
"in a break" or not. That is, the `VAL` register will report 16
samples equivalent to the last UART bit received (which is 0 when in
a break, and 1 for a stop bit on any valid transmission). We make the
assumption here that break is being "held", rather than just being a
transient event. This logic is enabled through an `oversample-break`
property to make it an opt-in behaviour.

To better support flows that use break as a GPIO signal that can be
continually asserted, add an additional property that treats incoming
break events as a "toggle", such that one break event will cause the
UART to enter break until another event tells it to exit break. This is
again an opt-in property intended for testing edge-case software that
uses UART Break in this way.

Signed-off-by: Alex Jones <alex.jones@lowrisc.org>

rivos-eblot commented on hw/opentitan/ot_uart.c:
```diff
@@ -618,6 +648,8 @@ static Property ot_uart_properties[] = {
     DEFINE_PROP_STRING("clock-name", OtUARTState, clock_name),
     DEFINE_PROP_LINK("clock-src", OtUARTState, clock_src, TYPE_DEVICE,
                      DeviceState *),
+    DEFINE_PROP_BOOL("oversample_break", OtUARTState, oversample_break, false),
```
should be `"oversample-break"` to match other properties in UART

rivos-eblot commented on hw/opentitan/ot_uart.c:
```diff
@@ -686,8 +719,11 @@ static void ot_uart_realize(DeviceState *dev, Error **errp)
     fifo8_create(&s->tx_fifo, OT_UART_TX_FIFO_SIZE);
     fifo8_create(&s->rx_fifo, OT_UART_RX_FIFO_SIZE);
 
+    s->in_break = false;
```
I do not think this is needed nor desirable in the realize function.
Everything is already zeroified.

However shouldn't it be reset in the reset_enter handler?

AlexJones0 replied:
That is a good point, but given that `in_break` is being used to signify a break condition held the UART RX, I think this should survive a reset? At least, the current ROM_EXT rescue SW basically uses it like a strap and assumes we can hold UART RX in break and reset the chip to enter Rescue boot services.

rivos-eblot replied:
It needs to be the same as the real HW. It is weird that the break condition survives a HW reset (sorry no time to dig into the RTL at the moment).

Is it not possible to evaluate the "break" condition when the UART goes out of reset, i.e. in `reset_exit`?

AlexJones0 replied:
Perhaps I am not explaining well - QEMU only supports the notion of break events - i.e. no notion of break being held or not. So you can receive a break event but that does not mean that the UART RX signal is being held in break by whatever is talking to QEMU. Break is not continually sent, it is a one-and-done event in QEMU.

Unfortunately the ROM_EXT rescue SW relies on being able to use UART RX like a SW strap by asserting (and de-asserting) break conditions on UART RX whilst rebooting/resetting OpenTitan. To support this functionality I have added the `oversample_break` property which will essentially "remember" whether the RX pin is in break or not. The analogue in real HW would be the UART coming out of reset and sampling the UART RX pin again and seeing the break almost immediately (*maybe this means on reset_exit if `in_break` I should fire the interrupt again?*).

AFAIK there is no easy way to evaluate the "break" condition when the UART goes out of reset because the UART CharDev only supports sending/receiving characters, or doing a break event via the QEMU monitor. Unless I build a protocol on top of the UART CharDev (not ideal) there isn't any way to poll whatever's talking to OpenTitan to ask "am I still in a break?"

rivos-eblot replied:
Ok, I finally got it. Thanks for the details. It might be worth adding (part of) this comment into the code itself to explain why this value is not reset, which is unusual.

AlexJones0 replied:
> Ok, I finally got it. 

Ideally in my opinion the ROM_EXT shouldn't be using UART RX like this at all (I don't know the original motivation, but I don't see why it could not just use and poll UART RX as a GPIO instead) but at least this should be a relatively small work-around and lets us support Earlgrey rescue and ownership transfer flows. :sweat: 

> It might be worth adding (part of) this comment into the code itself to explain why this value is not reset, which is unusual.

Agreed, I will add this.

jwnrt reviewed (commented):
Please could you describe the breaking system in the UART markdown document since we're using it slightly strangely?

AlexJones0 committed:
[ot] docs/opentitan: uart.md: Add UART documentation

Separate out the UART documentation from the Earlgrey and Darjeeling
machine documentation files into its own seperate documentation file
(like the `spi_device.md`) and describe in more detail sending break
conditions via the QEMU monitor and using the properties to mock break
conditions in the oversampled `VAL` register for SW edge cases.

Signed-off-by: Alex Jones <alex.jones@lowrisc.org>

rivos-eblot commented on docs/opentitan/darjeeling.md:
```diff
@@ -305,16 +305,6 @@ To limit the reboot cyckes, use the `-global ot-rstmgr.fatal_reset=<N>` option,
-  options are not specific to OpenTitan emulation, but are useful to communicate over a UART.
-  Note that QEMU offers many `chardev` backends, please check QEMU documentation for details.
-
 ## Useful debugging options
```
I would leave `### UART` with a link to the actual UART documentation ("see doc..."), otherwise it is quite easy to miss the documentation for this specific device
